src/Disks/S3/DiskS3.h

```diff
@@ -139,6 +170,18 @@ class DiskS3 : public IDisk
     UInt64 reserved_bytes = 0;
     UInt64 reservation_count = 0;
     std::mutex reservation_mutex;
+
+    std::atomic<UInt64> revision_counter;
+    static constexpr UInt64 LATEST_REVISION = (static_cast<UInt64>(1)) << 63;
+    static constexpr UInt64 UNKNOWN_REVISION = 0;
+
+    /// File at path {metadata_path}/restore contains metadata restore information
+    const String restore_file_name = "restore";
+    /// The number of keys listed in one request (1000 is max value)
+    int list_object_keys_size;
+
+    /// Key has format: ../../r{revision}-{operation}
+    const re2::RE2 key_regexp {".*/r(\\d+)-(\\w+).*"};
```

Comment:
Should in match data files (`rXXX-file-NAME`) and operations `rXXX-rename`/`rXXX-harlink` and only it?

Reply:
Yes, it's enough.
For S3 objects that don't' match regex, we can assume that is an old-style file object and it has no revision (0)